### PR TITLE
prevent double dispose on unit test in OOP

### PR DESCRIPTION
**Pull request by heejaechang**
@333fred found an issue in unit test around OOP double disposing.

(https://ci.dot.net/job/dotnet_roslyn/view/Official%20Builds/job/dev15.5.x/job/windows_release_unit32/46/testReport/junit/Microsoft.CodeAnalysis.Editor.UnitTests.ReferenceHighlighting/VisualBasicReferenceHighlightingTests/TestAccessor1/)

I haven't seen this for actual VS. so wondering whether this is unit test specific issue since OOP in unit test uses mock service hub (http://source.roslyn.io/#Roslyn.Services.Test.Utilities/Remote/InProcRemostHostClient.cs,19) I created.

still made the code to guard against double dispose in product as well but made it to crash in debug.

...

found out why it is happening in unit test. it only happens in unit test due to how WrappedStream (http://source.roslyn.io/#Roslyn.Services.Test.Utilities/Remote/InProcRemostHostClient.cs,239) is tied to service.

### CODE REVIEW

**Comment by heejaechang** [comment]
@dotnet/roslyn-analysis @dotnet/roslyn-ide @333fred can you take a look?

**Comment by heejaechang** [comment]
@sharwell @dotnet/roslyn-analysis @dotnet/roslyn-ide can you take a look?

**Comment by heejaechang** [comment]
retest windows_coreclr_test_prtest please

File: src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs [hunk 1]
```diff
@@ -42,29 +42,12 @@ internal abstract class ServiceHubServiceBase : IDisposable
...
-            Rpc.JsonSerializer.Converters.Add(AggregateJsonConverter.Instance);
-
-            Rpc.Disconnected += OnRpcDisconnected;
-        }
+        private bool _disposed;
 
         protected ServiceHubServiceBase(IServiceProvider serviceProvider, Stream stream)
         {
             InstanceId = Interlocked.Add(ref s_instanceId, 1);
+            _disposed = false;
```
Comment (sharwell) [inline src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs]: :bulb: This is unnecessary.

File: src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs [hunk 2]
```diff
@@ -130,6 +113,15 @@ public virtual void Initialize(PinnedSolutionInfo info)
...
+            {
+                // guard us from double disposing. this can happen in unit test
+                // due to how we create test mock service hub stream that tied to
+                // remote host service
+                return;
+            }
+
+            _disposed = true;
             Rpc.Dispose();
             _shutdownCancellationSource.Dispose();
```
Comment (sharwell) [inline src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs]: :memo: The line below this says `Dispose(false)`, but it should be `Dispose(true)`

File: src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs (hunk 1 above)
Comment (heejaechang) [reply src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs]: I know, but want to be explicit

File: src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs (hunk 2 above)
Comment (heejaechang) [reply src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs]: sure

**Comment by heejaechang** [comment]
@Pilchie @MattGertz forgot to ask for approval for preview 2? can you take a look?

**Pull request merged**